jobserver/models/core.py

```diff
@@ -1363,21 +1363,42 @@ def get_action_status_lut(self, backend=None):
 
         We need to get the latest status for each action run inside this
         Workspace.
-        """
-        jobs = Job.objects.filter(job_request__workspace=self)
-
-        if backend:
-            jobs = jobs.filter(job_request__backend__slug=backend)
-
-        # get all known actions
-        actions = set(jobs.values_list("action", flat=True))
 
-        action_status_lut = {}
-        for action in actions:
-            # get the latest status for an action
-            job = jobs.filter(action=action).order_by("-created_at").first()
-            action_status_lut[action] = job.status
-        return action_status_lut
+        We're doing this in raw SQL so we can use a CTE to get the latest
+        actions, and then join that with the Jobs table to get the status of
+        those actions.
+        """
+        sql = """
+        WITH latest_actions AS (
+          SELECT
+            job.action AS name,
+            max(job.created_at) AS latest
+          FROM
+            jobserver_job job
+            INNER JOIN jobserver_jobrequest job_request ON (job.job_request_id = job_request.id)
+            INNER JOIN jobserver_workspace workspace ON (job_request.workspace_id = workspace.id)
+          WHERE workspace.name = %s
+          GROUP BY action
+        )
+        SELECT
+          latest_actions.name AS name,
+          job.status AS status
+        FROM
+          latest_actions
+          INNER JOIN jobserver_job job ON (
+            latest_actions.name = job.action
+            AND
+            latest_actions.latest = job.created_at
```

Comment:
Joining back into one of the tables also referenced within your CTE feels like a bit of a smell to me. Did you evaluation the performance of  this approach (aggregate with max in the CTE, join back to source table on result of aggregation) to using `ROW_NUMBER()` within the CTE and the selecting the row from the CTE with the lowest row number? My instinct would be to prefer the latter rather but that might be a mssqlism rather than anything that matters in pgsql.

Reply:
```sql
WITH latest_actions AS (
  SELECT
    job.action as name,
    jon.status as status,
    row_number() OVER (PARTITION BY job.action ORDER BY job.created_at DESC) as row_num
  FROM 
    jobserver_job job
  INNER JOIN jobserver_jobrequest job_request ON (job.job_request_id = job_request.id)
  INNER JOIN jobserver_workspace workspace ON (job_request.workspace_id = workspace.id)
  WHERE workspace.name = %s
)
SELECT 
  name,
  status
FROM cte
WHERE row_num = 1
```

Reply:
I hadn't even thought to consider window functions!  And your version is nearly twice as fast on my local machine 👌